Adds @compound macro

LalitChauhan56 opened the pull request:
[WIP] Extension to this [PR](https://github.com/LalitChauhan56/TestPackage.jl/pull/9) which creates the `@compound` macro with the [updated input.](https://github.com/LalitChauhan56/TestPackage.jl/pull/9#issuecomment-1629558885). Also adds the `iscompound()` function to check if a species is a compound or not. 

```
julia>@variables t
@parameters k
@species C(t) H(t) O(t)

julia>@compound C6H12O2(t) 6C 12H 2O 
3-element Vector{Num}:
  6C(t)
 12H(t)
  2O(t)

julia> iscompound(C6H12O2)
true
```

isaacsas commented:
Can you add a corresponding tests file with some tests?

TorkelE commented:
Could you do a test that checks that the components of a compound is the expected one? Like that H2O indeed consist of 2 H and 1 O?

Also, for the test file to be run, you need to add it to the runtests.jl file. You could add it to this section:
```
    # Runs various miscellaneous tests.
    @time @safetestset "API" begin include("miscellaneous_tests/api.jl") end
    @time @safetestset "Symbolic Stoichiometry" begin include("miscellaneous_tests/symbolic_stoichiometry.jl") end
    @time @safetestset "Events" begin include("miscellaneous_tests/events.jl") end
    @time @safetestset "Units" begin include("miscellaneous_tests/units.jl") end
```
Maybe like:
```
    # Runs various miscellaneous tests.
    @time @safetestset "API" begin include("miscellaneous_tests/api.jl") end
    @time @safetestset "Symbolic Stoichiometry" begin include("miscellaneous_tests/symbolic_stoichiometry.jl") end
    @time @safetestset "Events" begin include("miscellaneous_tests/events.jl") end
    @time @safetestset "Compound species" begin include("miscellaneous_tests/compound_macro.jl") end
    @time @safetestset "Units" begin include("miscellaneous_tests/units.jl") end
```

I'd probably also move it to misc tests (?) that it does not necessarily require the DSL. 

Otherwise looks good :)

LalitChauhan56 commented:
From what I understand the `CompoundSpecies` struct is not being imported when the tests run? I included the compound.jl file using `include("compound.jl")` and exported the `@compound` macro along with the `iscompound` and `components` functions. Is there something else I am missing?

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,59 @@
+    escaped_setmetadata_expr = esc(setmetadata_expr)
+
+    # Construct the array from the remaining arguments
+    arr = Expr(:vect,(arr_expr)...)
```
Can you store the component information in a more easy to query manner. You could either have two metadata, `CompoundComponent` and `CompoundCoefficients` or have `CompoundComponent` return a vector of pairs of the form `[A => 3, B => 1, C => 2]` for A3BC2?

isaacsas replied:
@TorkelE and I spoke, and I think we prefer the approach of storing the symbolic components and their coefficients separately.

LalitChauhan56 replied:
Okay, will implement this in the next commit.

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,59 @@
+    species_arg = species_expr.args[2]
+
+    # Construct the expressions that define the species
+    species_expr = Expr(:macrocall, Symbol("@species"), LineNumberNode(0), Expr(:call, species_name, species_arg))
```
Does this support interpolation of previously defined species? i.e.
```julia
@variables t
@species A(t) B(t)
@compound A2B3 2*($A) 3*($B)
```
to use the existing A and B symbolic variables instead of creating new ones? The reason to support this is that there may be extra metadata attached to predefined species we want to preserve and keep in the compound.
```

LalitChauhan56 replied:
Before defining a compound the constituent species must be explicitly declared. The `@compound` macro only creates a single new species which is the compound itself. So if I use a previously declared species with the macro, it is not created again and so the metadata is preserved. 

So in this case, any metadata previously attached to species `A` or `B` will be accessible even after it has been used with the `@compound` macro.

isaacsas commented on test/miscellaneous_tests/compound_macro.jl:
```diff
@@ -0,0 +1,23 @@
+    @compound H2O(t) 1H 2O
+    @test typeof(H2O) == Num
+    arr_components = [1H ,2O]
+    @test all((string(c1) == string(c2) for (c1, c2) = zip(components(H2O), arr_components)))
```
You should test equality of the numeric coefficients and the symbolic variables separately. For symbolic variables you can use `isequal(A,B)` to test if they are the same symbolic. You shouldn't ever be converting to strings when working with symbolic variables except for display reasons.

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,59 @@
+    getmetadata(s, CompoundSpecies, false)
+end
+
+function components(s)
```
Add a dispatch if `s` is a `Num` that unwraps?

isaacsas commented:
You probably need to use `Catalyst.CompoundSpecies`. Take a look at the `@species` macro definition.

TorkelE commented:
Ok, so some specifications to this.

1) Storage. For each compound, let's store 1 vector `components` and ` vector `coefficients`. Here, e.g. for
```julia
@variables t
@species C(t) H(t) O(t)
@compound C6H12O2(t) 6C 12H 2O 
```
we would store `components` as `[C, H, O]`, where `C` is the `species` `C`, and `coefficients` as `[6, 12, 2]`. Create a test and ensure that this works:
```julia
let
  @variables t
  @species C(t) H(t) O(t)
  @compound C6H12O2(t) 6C 12H 2O 

  @test isequal([C, H, O], components(C6H12O2))
  @test isequal([6, 12, 2], coefficients(C6H12O2))
end
```
Here, `components` and `coefficients` are getter functions that retrieve the corresponding fields.

2) It would be good if the `@compound` command threw an error if one of the components where not described. E.g. make a test to ensure that
```julia
let
  @variables t
  @species C(t) H(t)
  @compound C6H12O2(t) 6C 12H 2O 
end
```
throws an error.

3) Nested components should work. E.g.
```julia
  @variables t
  @species C(t) H(t) O
  @compound OH O H
  @compound C3H5OH3 3C 5H 3OH

  @test isequal(components(components(C3H5OH3)[3]), [O, H])
  @test isequal(coefficients(components(C3H5OH3)[3]), [1, 1])
```

4) Interpolation. One should be able to do e.g.
```julia
@variables t
@species C(t) H(t) O(t)
s = C
@compound C6H12O2(t) 6$s 12H 2O 
```
and get the same result as 
```julia
@variables t
@species C(t) H(t) O(t)
@compound C6H12O2(t) 6C 12H 2O 
```

I will try to write some proper test cases for you to add to your file, which you can use as specifications, when I get back to the office tomorrow.

LalitChauhan56 commented:
Understood. Will implement these in the following commits. Thank you for the detailed explanation :)

TorkelE commented:
Ok, here is a slightly more comprehensive test file you can use to ensure that the functionality works. If there are any oddities, just report it and I will have a look.
```
using Catalyst, Test

# Test base funcationality in two cases.
let 
    @variables t
    @species C(t) H(t) O(t)
    @compound C6H12O2(t) 6C 12H 2O 
    
    @test iscompound(C6H12O2)
    @test isspecies(C6H12O2)
    @test !iscompound(C)
    @test !iscompound(H)
    @test !iscompound(O)

    @test isequal([C, H, O], components(C6H12O2))
    @test isequal([6, 12, 2], coefficients(C6H12O2))
    @test isequal([C => 6, H => 12, O => 2], component_coefficients(C6H12O2)) # Maybe chaneg "component_coefficients" name soem time.
    @test all(.! iscompound.components(C6H12O2))    
end
let 
    @variables t
    @species O(t)
    @compound O2(t) 2O
    
    @test iscompound(O2)
    @test isspecies(O2)
    @test !iscompound(O)

    @test isequal([O], components(O2))
    @test isequal([2], coefficients(O2))
    @test isequal([O => 2], component_coefficients(O2)) # Maybe chaneg "component_coefficients" name soem time.
    @test all(.! iscompound.components(O2))    
end

# Checks that compounds cannot be created from non-existing species.
let
    @variables t
    @species C(t) H(t)
    @test_throws Exception @compound C6H12O2(t) 6C 12H 2O    # @test_throws might behave weirdly here, if you think this fails as it should but the test comes out starneg, just report and we will figure it out.
end
let
    @variables t
    @test_throws Exception @compound O2(t) 2O    # @test_throws might behave weirdly here, if you think this fails as it should but the test comes out starneg, just report and we will figure it out.
end

# Checks that nested components works as expected.
let 
    @variables t
    @species C(t) H(t) O
    @compound OH O H
    @compound C3H5OH3 3C 5H 3OH

    @test !iscompound(O)
    @test !iscompound(H)
    @test iscompound(OH)
    @test iscompound(C3H5OH3)
    
    @test !iscompound(components(C3H5OH3)[1])
    @test !iscompound(components(C3H5OH3)[2])
    @test iscompound(components(C3H5OH3)[3])

    @test isequal([C, H, OH], components(C3H5OH3))
    @test isequal([O, H], components(components(C3H5OH3)[3]))
    @test isequal([3, 5, 3], coefficients(C3H5OH3))
end

# Checks that interpolation works.
let 
    @variables t
    @species C(t) H(t) O(t)
    s = C
    @compound C6H12O2_1(t) 6$s 12H 2O 
    @compound C6H12O2_2(t) 6C 12H 2O 

    @test isequal(components(C6H12O2_1), components(C6H12O2_2))
    @test isequal(coefficients(C6H12O2_1), coefficients(C6H12O2_2))
end
let 
    @variables t
    @species C(t) H(t)
    @compound Cyclopentadiene 5C 6H
    C5H6 = Cyclopentadiene
    @compound C10H12 2$(C5H6)

    @test iscompound(C10H12)
    @test iscompound(components(C10H12)[1])
    
    @test isequal(components(C10H12)[1], C5H6)
    @test isequal(components(C10H12)[1], Cyclopentadiene)
    @test isequal(coefficients(C10H12)[1], 2)
end
```

LalitChauhan56 commented:
Thank you! Will test this and let you know.

LalitChauhan56 commented:
After numerous attempts of trying to store the components and coefficients separately within the macro itself, I ran into hygiene and scoping issues with the macro. As a workaround, I switched back to the previous approach and instead made changes to the retrieval functions `components` and `coefficients` themselves. They now retrieve the metadata array from the species and returns the needed values. All the tests pass now for the `components`, `coefficients` and `iscompond` functions. Commented out the interpolation tests as that is something I am working on right now, will test them once it is done.

Also, the `@test all(.! iscompound.components(C6H12O2))` type lines in the test file gave error `iscompound has no field components` (also added the error as comments in the test file) so instead I switched it with `@test all(!iscompound(i) for i in components(C6H12O2))` which does the same thing (correct me if I'm wrong).

LalitChauhan56 commented:
Removed `$` from the tests as mentioned above. All tests pass now.
@isaacsas @TorkelE

LalitChauhan56 commented:
Also, how can I fix the format check tests? :D

ErikQQY commented:
To fix the format check, you will need to run ```using JuliaFormatter, Catalyst; format(joinpath(dirname(pathof(Catalyst)), ".."))``` in the package developing environment before commit.

isaacsas commented:
Don't reformat in this PR. The style recently changed, so it will polute the PR by modifying like ever Catalyst file...

There should be a new style update in the near future that will also require a complete repo reformat, and we can rerun formatting at that time.

isaacsas commented:
Until then just ignore the format related failures.

isaacsas commented:
@LalitChauhan56 I put in a fix to force the older formatter version. If you want to merge/rebase master into your PR you can use @ErikQQY's instructions to format provided you make sure to install/use version 1.0.32 of the formatter. 

It should then only reformat files you've updated.

LalitChauhan56 commented:
Fixed. Thank you! @isaacsas @ErikQQY

isaacsas commented on test/miscellaneous_tests/compound_macro.jl:
```diff
@@ -0,0 +1,97 @@
+let
+    @variables t
+    @species C(t) H(t) O(t)
+    s = C
```
Does it work if `s = 2 C?`

isaacsas replied:
Likewise can one use an existing parameter or variable as a coefficient, i.e.
```julia
@variables t a(t)
@parameters b
@species A(t) B(t)
c = 3*b
@compound AaB6 = a*A 2*c
```

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,80 @@
+    MT.getmetadata(s, CompoundSpecies, false)
+end
+
+coefficients(s::Num) = coefficients(MT.value(s))
+function coefficients(s)
+    Z = MT.getmetadata(s, CompoundData)
+    coefficients = [Symbolics.arguments(Symbolics.value(element))[1] for element in Z]
+    return coefficients
+end
+
+components(s::Num) = components(MT.value(s))
+function components(s)
+    Z = MT.getmetadata(s, CompoundData)
+    comp = Num[]
+
+    for component in Z
+        # extract the variables using the get_variables() function
+        extracted = get_variables(component)
+
+        # append the extracted variables to the new array
+        append!(comp, extracted)
+    end
+
+    return comp
+end
```
I'd prefer if these didn't allocate, and instead just returned an existing array.

Did you try looking at how the `@reaction_network` macro works? It calls code that can separate expressions to give the numerical coefficients. Maybe you can reuse that code / approach to split expressions within the macro instead of doing it in these function calls?

isaacsas reviewed (changes requested):
A couple high level comments.

isaacsas commented:
@LalitChauhan56 what’s the status on this?

LalitChauhan56 commented:
@isaacsas The macro now allocates within itself and not using the functions. The coefficients and components arrays are made and attached within the macro itself. However, I am still having issues with interpolation 😅. Also alongside to this, I am working on adding reaction balancing for reactions involving compounds. Since, I don't think interpolation will have any major effect on the balancing part of the code once it is finalized.

LalitChauhan56 commented:
I haven't been able to put in as many hours in the last 2-3 days due to some interviews. Rest assured I will put in extra hours this week to get back on track.

LalitChauhan56 commented:
This adds the function to balance a `Reaction` using the `balance` function. The function takes in a `Reaction `object and returns a balanced `Reaction `object.

For example -
```
@parameters k
@variables t
@species C(t) H(t) O(t)
@compound CO2(t) 1C 2O
@compound H2O(t) 2H 1O
@compound C6H12O6(t) 6C 12H 6O 
@compound O2(t) 2O

rx = Reaction(k,[CO2,H2O],[C6H12O6,O2])
> k, CO2 + H2O --> C6H12O6 + O2

balanced_rx = balance(rx)
> k, 6*CO2 + 6*H2O --> C6H12O6 + 6*O2
```

I plan to add tests for this next.
Source : http://mathgene.usc.es/matlab-profs-quimica/reacciones.pdf

isaacsas commented on Project.toml:
```diff
@@ -10,6 +10,7 @@ Graphs = "86223c79-3864-5bf0-83f7-82e725a168b6"
 JumpProcesses = "ccbc3e58-028d-4f4c-8cd5-9ae44345cda5"
 LaTeXStrings = "b964fa9f-0449-5b57-a5c2-d3ea65f4040f"
 Latexify = "23fbe1c1-3f47-55db-b15f-69d7ec21a316"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
```
Remove it from the extras if you are adding it to the main deps.

isaacsas commented on docs/src/assets/Project.toml:
```diff
@@ -0,0 +1,33 @@
+[deps]
```
This file shouldn't be checked in.

isaacsas commented on docs/src/example_networks/hodgkin_huxley_example.md:
```diff
@@ -0,0 +1,8 @@
+# [Hodgkin Huxley Example](@id hodgkin_huxley_example)
```
What is this file and modifications?

LalitChauhan56 replied:
Sorry, used my old PC for this commit and it messed up some files because it wasn't up-to-date with the repository. Will fix it in the next commit.

isaacsas commented on docs/src/example_networks/smoluchowski_coagulation_equation.md:
```diff
@@ -133,4 +133,4 @@ For the **additive kernel** we find
 ## References
```
What is this file being updated?

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,180 @@
+    b[end] = 1
+    
+    # Compute x= A^-1 *b
+    x = inv(A)*b
```
You should never solve Ax = b via taking the inverse. You should use a linear system solver, in this case we probably want one that works in integer arithmetic to avoid rounding. Take a look at what we do to calculate nullspaces in Catalyst, and at the code in:

https://github.com/SciML/ModelingToolkit.jl/blob/master/src/structural_transformation/bareiss.jl

which is designed to do row reduction while preserving integer matrices.

isaacsas replied:
https://gregorygundersen.com/blog/2020/12/09/matrix-inversion/

LalitChauhan56 replied:
Understood! Thank you for the information.

isaacsas commented on src/compound.jl:
```diff
@@ -0,0 +1,180 @@
+    x = inv(A)*b
+
+    # Normalize the stoichiometric coefficients to be integers.
+    smallest_value = minimum(x[x .> 0])  
+    x = x / smallest_value
+
+    return round.(Int, x)   # rounding is due to potential minimal numerical inaccuracies
```
Use an integer algorithm to remove the need for this.

LalitChauhan56 commented:
From what I understand GCD is actually creating issues in some cases. For example - 
When balancing the reaction `Reaction(k,[CO2,H2O],[C6H12O6,O2])` 
the final stoich values come out to be 
```
4-element Vector{Float64}:
 1.0
 1.0
 0.16666666666666666
 1.0
```

I am using this as a [source](https://matrixcalc.org/slu.html#solve-using-Montante-method%28%7B%7B1,0,6,0,0%7D,%7B2,1,6,2,0%7D,%7B0,2,12,0,0%7D,%7B0,0,0,1,1%7D%7D%29) to get the answers. Here too, `x3 `comes out to be `1/6 = 0.16666....` which is rounded off to 0.2, so I think some amount of rounding will be required.

![image](https://github.com/SciML/Catalyst.jl/assets/73524110/b285f47e-af59-4cfd-9cb2-fa4d84f49d7e)

TorkelE commented:
Are you adding balance of equations and such stuff here as well?

Generally, it is better to restrict 1 PR to 1 function. E.g. do one PR to add the `@compound` functionality, and then another for balancing equations and what not.

isaacsas commented:
@LalitChauhan56 will try to take a closer look in the next day or two, but shouldn't there be a way to do this that avoids floating point? Also, is your back substitution step really doing backsub, or is it doing a floating point least squares solve? (Is the matrix non-square, if so I believe it is likely doing least squares?)

isaacsas commented:
Also, @TorkelE is right that it probably makes sense to do the balancing as a separate PR so we can merge the compound stuff now.

LalitChauhan56 commented:
Understood, I'll create a separate PR for the balancing part.

Also, yes there should be a way to avoid floating point numbers. I'll try to figure it out.

LalitChauhan56 commented:
@isaacsas This commit uses rational numbers and skips floating point numbers entirely.

Also, once you have reviewed this I can delete the balancing code from this PR so that the `@compound` part can be merged and then I can make a new PR for the balancing part as that depends on the `@compound` part to work.

isaacsas commented:
Move the balancing stuff out of this PR now so we can get this merged, and then we can review it in the new PR you open. That’ll be easier for us all I think. Thanks!

isaacsas commented:
It is ok if the new PR has the code for this PR too; once this is merged you can just merge master into your new PR and the redundant code will get dropped.

LalitChauhan56 commented:
@isaacsas I removed the balancing code from this PR. Also can you please take a look at the formatting test failing? I formatted the file using Julia Formatter 1.0.32 but it still fails.

isaacsas commented on src/Catalyst.jl:
```diff
@@ -92,4 +92,10 @@ include("latexify_recipes.jl")
+# for creating compounds
+include("compound.jl")
+export @compound
+export components, iscompound, coefficients, component_coefficients, balance, get_stoich
```
You are exporting non-existent functions here? (i.e. `balance` and `get_stoich`?) Also, let's not export `component_coefficients`.

isaacsas commented on test/miscellaneous_tests/compound_macro.jl:
```diff
@@ -0,0 +1,98 @@
+    @test isequal([6, 12, 2], coefficients(C6H12O2))
+    @test isequal([C => 6, H => 12, O => 2], component_coefficients(C6H12O2)) # Maybe chaneg "component_coefficients" name soem time.
+    @test all(!iscompound(i) for i in components(C6H12O2))
+    # Test threw exception
+    # Expression: all(.!(iscompound.components(C6H12O2)))
+    # type #iscompound has no field components
```
What are these commented lines about?

isaacsas commented:
Basically, I want to ensure someone can use this programmatically where coefficients and/or species may be stored in a variable.

LalitChauhan56 commented:
I added more tests for the functionalities you suggested.

Also, regarding [this](https://github.com/SciML/Catalyst.jl/pull/651#discussion_r1264704726) issue, I am unable to make this work. In cases where there is a coefficient in the value of the variable itself such as `S = 2C` , in the macro I am left with two arguments. First, which is the original coefficient defined when using the `@compound` macro and the second is the one which is present in the value of the variable (2 in this case). To simplify this I tried using the `simplify()` function but it does not help and `eval()` runs into scoping issues because the macro cannot access the previously defined species/variables (`'C' is not defined in this case`). `esc()` also does not work as I already need to escape the expressions to set the metadata for components and coefficients and I cannot escape the same variable twice. Is there a way to make it work? If yes, then should I make the changes here itself or open a new PR later once this is merged as this enables basic functionality for `@compound`?

Letting it work the way it currently does leaves me incorrect components and coefficients. For example -
```
  @variables t
  @species C(t) H(t) O(t)
  s = 2C
  @compound C6H12O2_1(t) 6s 12H 2O
    
  julia> components(C6H12O2_1)
  3-element Vector{Num}:
   2C(t)  #Should be C
    H(t)
    O(t)

  julia> coefficients(C6H12O2_1)
  3-element Vector{Int64}:
    6 #Should be 2*6 = 12
   12
    2
```

LalitChauhan56 commented:
However cases such as `S = C `and the ones you have mentioned [above](https://github.com/SciML/Catalyst.jl/pull/651#pullrequestreview-1559442408) work perfectly fine.

isaacsas commented:
That sounds ok. I think it is fine if the right most term in the product has to be just a variable and not a general expression. We just want to make sure the coefficients can be expressions (and eventually document all this behavior).

isaacsas commented:
Looks good, we can clean up later if needed. Thanks, nice work on this PR!!!

isaacsas commented:
I'm not sure about the formatter issue. We can figure it out later.

I'll be traveling the next two weeks, but try to look at your compound PR when I get some time. @TorkelE said he will also try to work with you on any issues and reviewing that PR.

The next step is for you to now update that PR to master (either rebase master onto it, or merge master into it -- I usually do the latter).

LalitChauhan56 commented:
Understood. Thank you for guiding me throughout the PR :) 

Also, I have updated the balancing PR.
